### DISCUSSION TITLE
Using multiple regressors

### DISCUSSION THREAD

**Discussion by carson36** [winedarksea/AutoTS#84] [Q&A]
Hi, 

Are there any additional examples for creating models with multiple regressors? I see the use of fake_regressor() in the docs, but from what I understand this creates a regression with random numbers? What is the best way to create a model with multiple regressors? Thanks for the help!

**Comment by winedarksea** [comment]
Regressors can be a bit confusing as the term gets reused a bit in slightly different ways.
Models like UnivariateRegression, RollingRegression, and WindowRegression are using dozens to thousands of regressors as the "X" feature in modeling. These are automatically generated from the time series and date part information. I think this is what you are asking about, and short answer, nothing you need to do.

The `future_regressor` is a special case where you have information you already know about the future with a high degree of certainty. Fake regressor shows an example of how the input data should look. Strictly speaking it doesn't have to be about the future, just a value consistent across time that you want to apply to those dates. An example is with the `create_lagged_regressor` which you can see in https://github.com/winedarksea/AutoTS/blob/master/production_example.py#L85

A good example of real world `future_regressor` is something like "store open hours" and "discounts" used with predicting sales. Since both store hours and discounts can be planned in advance they can thus be known with a high degree of certainty about the future, and passed in, historical data for training and future data for prediction. 

A troubling example would be using weather data. Historical weather data can be passed as a regressor, and then weather forecasts can be passed for the future of the regressor. The problem is the weather forecasts introduce inaccuracy that doesn't exist in the training data, and may accordingly introduce additional error into the forecasts which the model is not expecting. 

Does that help?

**Reply by carson36** [reply]
Oh ok, that makes sense. So if I wanted to add a categorical regressor to my data, for example discounted vs retail price sales, how would I go about doing that?

**Reply by winedarksea** [reply]
`future_regressor` only accepts numeric input data. So your feature would need to be something like a 0 / 1 flag for whether an item is on sale or not, or potentially the % discount for the item as the regressor ... or both. 

Passing it in is pretty straightforward. The data needs to be in the `wide` style, which means a pd.DatetimeIndex (the data for which the value applies) and then each column of the dataframe will be a regressor (pass as many as you like, or just one). This is where `fake_regressor` can come in handy - just copy the data shape it creates, but with your own real data.
Pass in the dates matching the historical data to `.fit()` and pass in the data matching the future to `.predict()` 

To see whether the models chosen used your regressor, try the `model.used_regressor_check` which if True means it was used. Actually, sometimes if True it just means Holidays was used, but we'll ignore that edge case...

**Reply by carson36** [reply]
Ok, thanks a lot. That is a lot clearer. One more question: my company has different customers on two accounts, retail and wholesale. The wholesale orders are at a discounted rate. The distributions of each category is different, which is why I don't want to model them lumped together. In this case, is there a way to create a categorical regressor for retail and wholesale, or is this not possible because I can't predict the proportion of wholesale:retail ahead of time? Would it be better to create two entirely different models for each category? Thanks!

**Reply by winedarksea** [reply]
It does sound to me like you should try modeling them separately. Assuming you have plenty of data, that would be the way I would do it. The value of aggregating the two series is if you don't have that much data for a product. If suddenly the data becomes intermittent (lots of zeroes in the history) after splitting it apart, only then would it be harder to forecast when separated.

Something you could also try is passing in 2 time series for training, the first being total quantity sold, the other being the % of sales in retail (or wholesale) of the total. So then you forecast the total quantity and percentage, and can multiple those numbers to get the sales by type. How well that works is going to depend on how stable the product mix is. If one day it is almost all retail sales, and the next wholesale and so on back and forth, that is going to be very difficult to predict compared to a case where the split is consistent.

Honestly, it is hard to say what is best, especially without seeing the data.
Try every approach and see what looks like is giving you the best results.

**Comment by carson36** [comment]
Thanks a  lot! This is really helpful
